### PR TITLE
Checkout integration tests for set or change addresses

### LINKED ISSUE #506: Checkout tests: test cases posting data
We need tests to test some checkout scenarios like providing (or choosing) shipping address, shipping method (if booth required) and billing address. I propose add following tests using client, and post data:
- [ ] user pass new valid shipping address - if is save in session storage
- [ ] user pass new valid shipping address - if is save in order object, after finish checkout
- [ ] user pass new valid shipping address - if is add to authorized user (ass default shipping address)
- [ ] user pass new valid billing address - if is save in session storage
- [ ] user pass new valid billing address - if is save in order object, after finish checkout
- [ ] user pass new valid billing address - if is add to authorized user (ass default billing address)
- [ ] user pass new valid shipping address, and use it as billing address - if is save in order
- [ ] user pass new valid shipping address, and use it as billing address - if is save in as default addresses
- [ ] user choose existing shipping address - it is save in session storage
- [ ] user choose existing shipping address - it is save in order
- [ ] user choose existing shipping address - it is save ass default shipping address
- [ ] user choose existing shipping address - if is not save as new address (duplicate)
- [ ] user choose existing billing address - if is save in session storage
- [ ] user choose existing billing address - if is save in order
- [ ] user choose existing billing address - if is save as default billing address
- [ ] user choose existing shipping address - if is not save as new address (duplicate)
- [ ] user choose existing shipping method - if is save in session storage
- [ ] user choose existing shipping method - if is save in order
- [ ] user choose existing shipping method, then change it - if new one is save in session storage
- [ ] user choose existing shipping method, then change it - if new one is save in order
- [ ] user choose existing shipping address, then change it - if new one is save in session storage
- [ ] user choose existing shipping address, then change it - if new one is save in order
- [ ] user choose existing shipping address, then change own address - if chosen is save as new in storage
- [ ] user choose existing shipping address, then change own address - if chosen is save as new in order
- [ ] user choose existing billing address, then change own address - if chosen is save as new in storage
- [ ] user choose existing billing address, then change own address - if chosen is save as new in order
- [ ] user choose shipping method, after checkout method it is removed - if chosen is save in order
- [ ] user without chosen shipping address tries to choose it as billing address - if is not save in storage
- [ ] unauthorized user provide valid email address in shipping step - if is save in session storage
- [ ] unauthorized user provide valid email address in shipping step - if is save in order
- [ ] unauthorized user provide valid email address in billing step - if is save in session storage
- [ ] unauthorized user provide valid email address in billing step - if is save in order
- [ ] authorized user change own email after checkout - if is not changed in order
- [ ] user pass new invalid shipping address
- [ ] user pass new invalid billing address

### PR DESCRIPTION
**Pull request by maciekjablonski**
Fixes #506

### CODE REVIEW

File: tests/test_checkout_integration.py [hunk 1]
```diff
@@ -192,6 +195,60 @@ def test_client_login(request_cart_with_item, client, admin_user):
     assert response.context['checkout'].cart.token == request_cart_with_item.token
 
 
+def test_unauthorized_email_is_saved_shipping(client, customer_user,  # pylint: disable=W0613
+                                              request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
+     unauthorized user provide valid email address in shipping step -
```
Comment (patrys) [inline tests/test_checkout_integration.py]: This is not a valid docstring

File: tests/test_checkout_integration.py [hunk 2]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
 
     summary_response = client.post(url, data=address_data, follow=True)
     assert summary_response.context['order'].voucher is None
+
+
+def test_user_pass_new_valid_shipping_address(authorized_client, customer_user,  # pylint: disable=W0613,R0914
+                                              request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
+     user pass new valid shipping address
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

File: tests/test_checkout_integration.py [hunk 3]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
...
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_shipping_address, exclude=['id']) == \
+           model_to_dict(order.shipping_address, exclude=['id'])
+
+
+def test_user_pass_new_valid_billing_address(authorized_client, customer_user,  # pylint: disable=W0613,R0914
+                                             request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
+     user pass new valid billing address
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

File: tests/test_checkout_integration.py [hunk 4]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
...
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_billing_address, exclude=['id']) == \
+           model_to_dict(order.billing_address, exclude=['id'])
+
+
+def test_user_choose_existing_shipping_address(authorized_client, billing_address, customer_user,  # pylint: disable=W0613,R0914
+                                               request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
+     user choose existing shipping address
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

File: tests/test_checkout_integration.py [hunk 5]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
...
+    assert order_dict == model_to_dict(order.shipping_address, exclude=['id'])
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_shipping_address, exclude=['id']) == \
+           model_to_dict(order.shipping_address, exclude=['id'])
+
+
+def test_user_choose_existing_billing_address(authorized_client, billing_address, customer_user,  # pylint: disable=R0914
+                                              request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

File: tests/test_checkout_integration.py [hunk 6]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
...
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_billing_address, exclude=['id']) ==\
+           model_to_dict(billing_address, exclude=['id'])
+
+
+def test_user_choose_existing_shipping_method(authorized_client, billing_address, customer_user,  # pylint: disable=R0914
+                                              request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """
+    user choose existing shipping method
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

File: tests/test_checkout_integration.py [hunk 7]
```diff
@@ -291,3 +348,424 @@ def test_voucher_invalid(client, request_cart_with_item, shipping_method, vouche
...
+
+    shipping_method.delete()
+    assert order.groups.all()[0].shipping_method_name == name
+
+
+def test_user_choose_existing_shipping_method_then_change_it(authorized_client, billing_address,  # pylint: disable=R0914
+                                                             customer_user, request_cart_with_item,  # pylint: disable=W0613
+                                                             shipping_method):
+    """
+    user choose existing shipping method, then change it
```
Comment (patrys) [inline tests/test_checkout_integration.py]: Not a valid docstring

**Comment by krzysztofwolski** [comment]
@maciekjablonski How is this going? I have found bug in address assigner

**Comment by artursmet** [comment]
@maciekjablonski Any estimation on this?

**Comment by artursmet** [comment]
@maciekjablonski Please rebase this branch with latest master

File: tests/test_checkout_integration.py [hunk 8]
```diff
@@ -306,3 +357,427 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+    address_data = {'address': 'shipping_address'}
+    summary_response = authorized_client.post(shipping_method_response.request['PATH_INFO'],
+                                              data=address_data, follow=True)
+    order = summary_response.context['order']
+    order_dict = model_to_dict(order.shipping_address, exclude=['id'])
+    assert order_dict == shipping_data
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_shipping_address, exclude=['id']) == \
+           model_to_dict(order.shipping_address, exclude=['id'])
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: Why don't you use order_dict? 
Don't use \

File: tests/test_checkout_integration.py [hunk 9]
```diff
@@ -306,3 +357,427 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
                                    follow=True, HTTP_REFERER=url)
     assert voucher_response.status_code == 200
     assert voucher_response.context['checkout'].voucher_code is None
+
+
+def test_user_pass_new_valid_shipping_address(authorized_client, customer_user,  # pylint: disable=W0613,R0914
+                                              request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """User pass new valid shipping address
+      - if is save in session storage
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: What do you mean?

File: tests/test_checkout_integration.py [hunk 10]
```diff
@@ -306,3 +357,427 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+
+    """
+
+    # Enter checkout
+    checkout_index = authorized_client.get(reverse('checkout:index'), follow=True)
+    # Checkout index redirects directly to shipping address step
+    shipping_address = authorized_client.get(checkout_index.request['PATH_INFO'])
+
+    # Enter shipping address data
+    shipping_data = {
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: Make it a fixture. You're using it multiple times.

File: tests/test_checkout_integration.py (hunk 9 above)
Comment (maciekjablonski) [inline tests/test_checkout_integration.py]: Sorry, it was my note during development for me. I'll clean it

File: tests/test_checkout_integration.py (hunk 10 above)
Comment (maciekjablonski) [inline tests/test_checkout_integration.py]: We use form data explicite

Comment (patrys) [inline tests/test_checkout_integration.py]: In National Assembly code I have fixtures with example address data.

Comment (maciekjablonski) [inline tests/test_checkout_integration.py]: @mociepka and @artursmet ?

Comment (mociepka) [inline tests/test_checkout_integration.py]: If we don't have address generators I prefer to repeat test data if necessary. Maybe we could parametrize this tests?

Comment (patrys) [inline tests/test_checkout_integration.py]: What's wrong with having a `valid_address` fixture? The address is not the point of this test.

**Comment by maciekjablonski** [comment]
@krzysztofwolski fixed

File: tests/test_checkout_integration.py [hunk 11]
```diff
@@ -215,10 +245,9 @@ def test_email_is_saved_in_order(authorized_client, billing_address, customer_us
     assert order.user_email == customer_user.email
 
 
-def test_voucher_invalid(client, request_cart_with_item, shipping_method, voucher):  # pylint: disable=W0613,R0914
-    """
-    Look: #549 #544
-    """
+def test_voucher_invalid(client, request_cart_with_item, shipping_method, valid_address, voucher):  # pylint: disable=W0613,R0914
+    """Look: #549 #544"""
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: We can make it even better: 
```
Check if:
- user is forced to review checkout if used voucher becomes invalid
- vouchers cannot be used past their limit
```

File: tests/test_checkout_integration.py [hunk 12]
```diff
@@ -306,3 +313,336 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+    assert name == smart_text(shipping_method.price_per_country.all()[0])
+
+    shipping_method.delete()
+    assert order.groups.all()[0].shipping_method_name == name
+
+
+def test_user_choose_existing_shipping_method_then_change_it(authorized_client, billing_address,  # pylint: disable=R0914
+                                                             customer_user, request_cart_with_item,  # pylint: disable=W0613
+                                                             shipping_method, valid_address):
+    """User choose existing shipping method, then change it
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: I don't think we need docstrings if it only contain function name.

File: tests/test_checkout_integration.py [hunk 13]
```diff
@@ -306,3 +313,336 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+    assert order_dict == model_to_dict(billing_address, exclude=['id'])
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_billing_address, exclude=['id']) ==\
+           model_to_dict(billing_address, exclude=['id'])
+
+
+def test_user_choose_existing_shipping_method(authorized_client, billing_address, customer_user,  # pylint: disable=R0914
+                                              request_cart_with_item, shipping_method, valid_address):  # pylint: disable=W0613
+    """User choose existing shipping method
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: I don't think we need docstrings if it only contain function name.

File: tests/test_checkout_integration.py [hunk 14]
```diff
@@ -306,3 +313,336 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+
+    default_shipping_address_dict = model_to_dict(order.shipping_address, exclude=['id'])
+    shipping_address_dict = model_to_dict(order.shipping_address, exclude=['id'])
+    assert default_shipping_address_dict == shipping_address_dict
+
+
+
+def test_user_choose_existing_billing_address(authorized_client, billing_address, customer_user,  # pylint: disable=R0914
+                                              request_cart_with_item, shipping_method, valid_address):  # pylint: disable=W0613
+    """User choose existing billing address
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: I don't think we need docstrings if it only contain function name.

File: tests/test_checkout_integration.py [hunk 15]
```diff
@@ -306,3 +313,336 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+    order_dict = model_to_dict(order.billing_address, exclude=['id'])
+    assert order_dict == valid_address
+
+    user = User.objects.get(pk=customer_user.pk)
+    assert model_to_dict(user.default_billing_address, exclude=['id']) == order_dict
+
+
+def test_user_choose_existing_shipping_address(authorized_client, billing_address, customer_user,  # pylint: disable=W0613,R0914
+                                               request_cart_with_item, shipping_method):  # pylint: disable=W0613
+    """User choose existing shipping address
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: I don't think we need docstrings if it only contain function name.

File: tests/test_checkout_integration.py [hunk 16]
```diff
@@ -306,3 +313,336 @@ def test_remove_voucher(client, request_cart_with_item, shipping_method, voucher
...
+
+    user = User.objects.get(pk=customer_user.pk)
+    default_shipping_address_dict = model_to_dict(user.default_shipping_address, exclude=['id'])
+    shipping_address_dict = model_to_dict(order.shipping_address, exclude=['id'])
+    assert default_shipping_address_dict == shipping_address_dict
+
+
+def test_user_pass_new_valid_billing_address(authorized_client, customer_user,  # pylint: disable=W0613,R0914
+                                             request_cart_with_item, shipping_method, valid_address):  # pylint: disable=W0613
+    """User pass new valid billing address
```
Comment (krzysztofwolski) [inline tests/test_checkout_integration.py]: I don't think we need docstrings if it only contain function name.

File: tests/test_checkout_integration.py [hunk 17]
```diff
@@ -180,11 +171,53 @@ def test_summary_without_shipping_method(request_cart_with_item, client, monkeyp
     assert get_redirect_location(response) == reverse('checkout:shipping-method')
 
 
+def test_unauthorized_email_is_saved_shipping(client, customer_user,  # pylint: disable=W0613
+                                              request_cart_with_item, shipping_method, valid_address):  # pylint: disable=W0613
+    """Unauthorized user provide valid email address in shipping step -
```
Comment (szewczykmira) [inline tests/test_checkout_integration.py]: `... user provides ...`

File: tests/test_checkout_integration.py [hunk 18]
```diff
@@ -180,11 +171,53 @@ def test_summary_without_shipping_method(request_cart_with_item, client, monkeyp
...
+    order = summary_response.context['order']
+    assert order.user_email == 'test@example.com'
+
+
 def test_email_is_saved_in_order(authorized_client, billing_address, customer_user,  # pylint: disable=R0914
                                  request_cart_with_item, shipping_method):
-    """
-    authorized user change own email after checkout - if is not changed in order
-    """
+    """Authorized user change own email after checkout - if is not changed in order"""
```
Comment (szewczykmira) [inline tests/test_checkout_integration.py]: `... user changes ...`

**Pull request closed without merging**